src/algorithms/arc_swap.rs

```diff
@@ -103,164 +101,204 @@
-                            .unwrap()
-                            .iter()
-                            .map(|(neighbor2, edge_weight)| {
-                                let part = partition[neighbor2].load(Ordering::Relaxed);
-                                if part == neighbor_part {
-                                    -*edge_weight
-                                } else if part == target_part {
-                                    *edge_weight
-                                } else {
-                                    0
-                                }
-                            })
-                            .sum()
-                    })
-                    .max()
-                    .unwrap();
-                if 0 < neighbor_gain {
-                    cut.push(neighbor);
-                }
+        for (neighbor, _edge_weight) in adjacency.outer_view(moved_vertex).unwrap().iter() {
+            let neighbor_part = partition[neighbor].load(Ordering::Relaxed);
+            let neighbor_gain: i64 = (0..part_count)
+                .filter(|target_part| *target_part != neighbor_part)
+                .map(|target_part| {
+                    adjacency
+                        .outer_view(neighbor)
+                        .unwrap()
+                        .iter()
+                        .map(|(neighbor2, edge_weight)| {
+                            let part = partition[neighbor2].load(Ordering::Relaxed);
+                            if part == neighbor_part {
+                                -*edge_weight
+                            } else if part == target_part {
+                                *edge_weight
+                            } else {
+                                0
+                            }
+                        })
+                        .sum()
+                })
+                .max()
+                .unwrap();
+            if 0 < neighbor_gain {
+                cut.push(neighbor);
             }
+        }
 
-            true
-        };
+        true
+    };
 
-    let mut metadata =
-        partition
-            .par_iter()
+    let mut metadata = Metadata::default();
+    // Thread-local maximum of the weight of each part.
+    // Each thread can move vertices as they wish as long as their local
+    // `part_weights` remains lower than this array, element-by-element.
+    // While this can be over-restrictive, it removes a RW lock and still
+    // ensures the imbalance stays within bounds.
+    let mut thread_max_pws = vec![W::zero(); part_count];
+    loop {
+        metadata.pass_count += 1;
+
+        // `part_weights` changes at the end of each pass, so `thread_max_pws`
+        // needs to be updated here.
+        for (max_pw, pw) in thread_max_pws.iter_mut().zip(&part_weights) {
+            *max_pw = *pw
+                + W::from_f64((max_part_weight - *pw).to_f64().unwrap() / thread_count as f64)
+                    .unwrap();
+        }
+
```

Comment:
I am still not convinced by this kind of over constraint.

Reply:
I agree this is overly restrictive. The problem is that there is no way to predict how much and where threads will move weights before the pass starts.

To relax this constraint, we would need either:
1. synchronize the `part_weights` array and greatly reduce the performance,
2. do some prior work to make it so threads don't all move vertices to one part, somehow?

Please note that, since `thread_max_pws` is updated before each pass, so any imbalance-breaking move can be retried at a later pass if necessary.